akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -491,3 +495,51 @@ private[akka] final case class MapAsyncUnordered[In, Out](parallelism: Int, f: I
     if (todo > 0) ctx.absorbTermination()
     else ctx.finish()
 }
+
+/**
+ * INTERNAL API
+ */
+private[akka] final case class DebugLogging[T](name: String, extract: T ⇒ Any, userLogAdapter: Option[LoggingAdapter]) extends PushStage[T, T] {
+
+  private final val DefaultLogLevels = LogLevels(onElement = Logging.DebugLevel, onFinish = Logging.DebugLevel, onFailure = Logging.ErrorLevel)
+
+  private var log: LoggingAdapter = _
+  private var logLevels: OperationAttributes.LogLevels = _
+
+  override def onPush(elem: T, ctx: Context[T]): SyncDirective = {
+    initLogger(ctx)
+
+    log.log(logLevels.onElement, "[{}] Element: {}", name, extract(elem))
+    ctx.push(elem)
+  }
+
+  override def onUpstreamFailure(cause: Throwable, ctx: Context[T]): TerminationDirective = {
+    initLogger(ctx)
+
+    // TODO should be able to determine who my upstream is and give it's name here?
+    logLevels.onFailure match {
+      case Logging.ErrorLevel ⇒ log.error(cause, "[{}] Upstream failed.", name)
+      case level              ⇒ log.log(level, "[{}] Upstream failed, cause: {}: {}", name, Logging.simpleName(cause.getClass), cause.getMessage)
+    }
+    super.onUpstreamFailure(cause, ctx)
+  }
+
+  override def onUpstreamFinish(ctx: Context[T]): TerminationDirective = {
+    initLogger(ctx)
+
+    // TODO should be able to determine who my upstream is and give it's name here?
+    log.log(logLevels.onFinish, "[{}] Upstream finished.", name)
+    super.onUpstreamFinish(ctx)
+  }
+
+  /** Idempotent initialization of logger and log levels */
+  def initLogger(ctx: Context[T]): Unit =
+    if (log == null) {
+      logLevels = ctx.attributes.logLevels.getOrElse(DefaultLogLevels)
+
+      log = userLogAdapter getOrElse {
+        val sys = ctx.materializer.asInstanceOf[ActorFlowMaterializerImpl].system
+        Logging(sys, this.getClass)
```

Comment:
Is it possible to amend that to something more public/stable, e.g. "akka.stream.Log"?
It is used in slf4j config.